zerver/lib/mention.py

```diff
@@ -269,15 +270,51 @@ def message_has_topic_wildcards(self) -> bool:
         self.user_group_name_info: dict[str, NamedUserGroup] = {}
-        self.user_group_members: dict[int, set[int]] = {}
+        self.user_group_members: dict[int, set[int]] = defaultdict(set)
         user_group_names = possible_user_group_mentions(content)
         if user_group_names:
-            for group in NamedUserGroup.objects.filter(
+            named_user_groups = NamedUserGroup.objects.filter(
                 realm_id=realm_id, name__in=user_group_names
-            ):
-                self.user_group_name_info[group.name.lower()] = group
-                self.user_group_members[group.id] = set(
-                    get_recursive_group_members(group).values_list("id", flat=True)
+            )
+            # If there are no matching user_group_names (e.g non-existent user_group_names),
+            # then no need to execute the relatively expensive recursive CTE query
+            # triggered by get_root_id_annotated_recursive_subgroups_for_groups
+            if len(named_user_groups) == 0:
+                return
+
+            self.user_group_name_info = {group.name.lower(): group for group in named_user_groups}
+
+            # Here we fetch membership for non-deactivated groups in a
+            # single, efficient bulk query. Deactivated group mentions
+            # are non-operative, so we don't need to waste resources
+            # collecting membership data for them. Further possible
+            # optimizations include, in order of value:
+            #
+            # - Skipping fetching membership data for silent mention syntax.
+            # - Filtering groups the current user doesn't have permission to mention
+            #   from what we fetch membership for. This requires care to avoid race bugs
+            #   if the user's permissions change while the message is being sent.
+            hierarchial_groups = get_root_id_annotated_recursive_subgroups_for_groups(
+                [group.id for group in named_user_groups if not group.deactivated], realm_id
+            ).prefetch_related(
+                # Here we use the `to_attr` feature of Prefetch to
+                # fetch the queryset of active members only.
+                Prefetch(
+                    "direct_members",
+                    queryset=UserProfile.objects.filter(is_active=True).only("id"),
+                    to_attr="active_direct_members",
+                )
+            )
+
+            # Possible BUG: What is root_id if there are multiple
+            # mentioned groups that contain a given group as a
+            # subgroup?
+            for group in hierarchial_groups:
+                self.user_group_members[group.root_id].update(  # type: ignore[attr-defined] # root_id is an annotated attribute.
+                    [
+                        member.id
+                        for member in group.active_direct_members  # type: ignore[attr-defined] # custom attribute.
+                    ]
```

Comment:
This can be simplified even more (and use one fewer query):

```python
for group_id, member_id in (
    get_root_id_annotated_recursive_subgroups_for_groups(
        [group.id for group in named_user_groups if not group.deactivated], realm_id
    )
    .filter(direct_members__is_active=True)
    .values_list("root_id", "direct_members")
):
    self.user_group_members[group_id].add(member_id)
```

Reply:
Yes, this works !